### ISSUE TITLE
Mojarra HighCost of its Initialization Strategy - Classpath scanning - should be improved and more configurable

### ISSUE THREAD

**Issue by 99sono** [javaserverfaces/mojarra#4298]
Hi,

I am currently trying to optimize the deployment cost of a specific application.
I am looking at all layers of the application that induce deployment cost (e.g. eclipselink was one bottleneck see https://issues.jboss.org/browse/WFLY-9408)

But now I stumbled upon a bottleneck on Mojarra.

In an abitrary WAR application that I am looking at, on average, the cost of starting up Mojarra during the "ServerClontext.contextInitialized" call is about 4.4 seconds.
I have taken several thread dumps of the startup cost of this applicatin, and when Mojarra starts initializing, invariable there are two Classes used by the  `com.sun.faces.config.ConfigManager` that pop up.

# LoadUrls on MetaInfFacesConfigResourceProvider
The first culprit for a heavy bootstrapping cost is the loadURLs method of the `com.sun.faces.config.configprovider.MetaInfFacesConfigResourceProvider`.

This load urls method is essentially responsible for hunting in the classpath of the deployed application for all of the faces-config.xml files.
The implementation of the method here is very interesting because it has two very relavent steps of completely different performance.

So if we look at the implementation of the method we have:

```
 private Collection<URI> loadURLs(ServletContext context) throws IOException {

        Set<URI> urls = new HashSet<URI>();
        try {
            for (Enumeration<URL> e = Util.getCurrentLoader(this).getResources(META_INF_RESOURCES); e.hasMoreElements();) {
                String urlString = e.nextElement().toExternalForm();
                urlString = urlString.replaceAll(" ", "%20");                
                urls.add(new URI(urlString));
            }
            URL [] urlArray = Classpath.search("META-INF/", FACES_CONFIG_EXTENSION);
            for (URL cur : urlArray) {
                String urlString = cur.toExternalForm();
                urlString = urlString.replaceAll(" ", "%20");
                urls.add(new URI(urlString));
            }
            // special case for finding taglib files in WEB-INF/classes/META-INF
            Set paths = context.getResourcePaths(WEB_INF_CLASSES);
            if (paths != null) {
                for (Object path : paths) {
                    String p = path.toString();
                    if (p.endsWith(FACES_CONFIG_EXTENSION)) {
                        String urlString = context.getResource(p).toExternalForm();
                        urlString = urlString.replaceAll(" ", "%20");
                        urls.add(new URI(urlString));
                    }
                }
            }
        } catch (URISyntaxException ex) {
            throw new IOException(ex);
        }
        return urls;
        
    }
```

## Fast code
The first step, at least on wildfly, I would have to benchmark it on weblogic...
`Enumeration<URL> e = Util.getCurrentLoader(this).getResources(META_INF_RESOURCES); e.hasMoreElements();`

this is Blazing fast. I can put a break point on this line, and the step next comes out with result instantaneously.
And best part of it, is has found all of the faces-config.xml on the application. All of them, because in this application there are no faces-config.xml files put at strange positions.

## Time burner step
The second step starts at this line of cone, 
` Classpath.search("META-INF/", FACES_CONFIG_EXTENSION)`

And this is not fast, not fast at all.
In fact, for the duration of 2.2seconds I get thread dumps of the form.

> SAMPLE 31 and 32 exactly the same:
> at java.io.WinNTFileSystem.getBooleanAttributes(Native Method) 
> at java.io.File.isDirectory(Unknown Source) 
> at org.jboss.vfs.spi.RootFileSystem.isDirectory(RootFileSystem.java:107) 
> at org.jboss.vfs.VirtualFile.isDirectory(VirtualFile.java:291) 
> at org.jboss.vfs.VirtualFile.openStream(VirtualFile.java:305) 
> at org.jboss.vfs.VirtualJarInputStream.openCurrent(VirtualJarInputStream.java:223) 
> at org.jboss.vfs.VirtualJarInputStream.getNextJarEntry(VirtualJarInputStream.java:109) 
> at org.jboss.vfs.VirtualJarInputStream.getNextEntry(VirtualJarInputStream.java:80) 
> at com.sun.faces.facelets.util.Classpath.searchFromURL(Classpath.java:222) 
> at com.sun.faces.facelets.util.Classpath.searchFromURL(Classpath.java:242) 
> at com.sun.faces.facelets.util.Classpath.search(Classpath.java:156) 
> at com.sun.faces.facelets.util.Classpath.search(Classpath.java:110) 
> 
> ---- Search for the MetaInfFiles (perhaps the faces-config.xml)
> at com.sun.faces.config.configprovider.MetaInfFacesConfigResourceProvider.loadURLs(MetaInfFacesConfigResourceProvider.java:166) 
> at com.sun.faces.config.configprovider.MetaInfFacesConfigResourceProvider.getResources(MetaInfFacesConfigResourceProvider.java:110) 
> at com.sun.faces.config.ConfigManager$URITask.call(ConfigManager.java:1362) 
> at com.sun.faces.config.ConfigManager$URITask.call(ConfigManager.java:1331) 
> at java.util.concurrent.FutureTask.run(Unknown Source) 
> 
> ---------------------------------------------------------------------
> -- NOTE: 
> -- Here mojarra is doing getConfigDocuments (same as sample 31 and 32)
> ---------------------------------------------------------------------
> at com.sun.faces.config.ConfigManager.getConfigDocuments(ConfigManager.java:742) 
> at com.sun.faces.config.ConfigManager.initialize(ConfigManager.java:353) 
> at com.sun.faces.config.ConfigureListener.contextInitialized(ConfigureListener.java:227) 
> 
> at io.undertow.servlet.core.ApplicationListeners.contextInitialized(ApplicationListeners.java:187) 
> at io.undertow.servlet.core.DeploymentManagerImpl$1.call(DeploymentManagerImpl.java:200) 
> at io.undertow.servlet.core.DeploymentManagerImpl$1.call(DeploymentManagerImpl.java:171) 
> at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:42) 
> at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
> at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 

And actualy, after these two seconds of classpath search, I get no results, because there no files suffixed as faces-config.xml at any strange locations.

So the first point for this bug report is.
In order to gain well needed deployment time and reduce the overall cost of the Mojarra initalization, it would be IDEAL, that through some sort of configuration parameter (e.g. a system property or any other type of configuration deemed appropriate by the mojarra team) that I am allowed to spefieid:
I do not want the ClasspathSearch for files withing META-INF/ at some strage depth with faces-config.xml to be searched for.
Essentially saying: my application does not do strange things, I only want to find faces-config.xml files at the META-INF depth and then I only need the first approach to get these config files and I can disable the slow approach.

I think this is direly needed here. Really critical.
It would already reduce by half the total time I am seing as a cost to initialize mojarra on wildfly.

# The hunt for .taglib.xml files

The second problem I have detected in the mojarra initialization is related with the 
`com.sun.faces.config.configprovider.MetaInfFaceletTaglibraryConfigProvider`

The problem is very similar to the problem described above with the hunt for faces-config.xml files, but with a bigger problem. There is not fast approach here, there is only the slow and expensive classpath search approach.

## Problematic code snippet

   ```
 public Collection<URI> getResources(ServletContext context) {

        try {
            
    public Collection<URI> getResources(ServletContext context) {

        try {
            URL[] externalTaglibUrls = Classpath.search(Util.getCurrentLoader(this),
                                          "META-INF/",
                                          SUFFIX);
            URL[] builtInTaglibUrls = new URL[BUILT_IN_TAGLIB_XML_FILES.length];
            ClassLoader runtimeClassLoader = this.getClass().getClassLoader();
            for (int i = 0; i < BUILT_IN_TAGLIB_XML_FILES.length; i++) {
                builtInTaglibUrls[i] = runtimeClassLoader.getResource(BUILT_IN_TAGLIB_XML_FILES[i]);
            }
            URL[] urls = new URL[externalTaglibUrls.length + builtInTaglibUrls.length];
            System.arraycopy(externalTaglibUrls, 0, urls, 0, externalTaglibUrls.length);
            System.arraycopy(builtInTaglibUrls, 0, urls, externalTaglibUrls.length, builtInTaglibUrls.length);
            // perform some 'correctness' checking.  If the user has
            // removed the FaceletViewHandler from their configuration,
            // but has left the jsf-facelets.jar in the classpath, we
            // need to ignore the default configuration resouces from
            // that JAR.
            List<URI> urlsList = pruneURLs(urls);

            // special case for finding taglib files in WEB-INF/classes/META-INF
            Set paths = context.getResourcePaths(WEB_INF_CLASSES);
            if (paths != null) {
                for (Object path : paths) {
                    String p = path.toString();
                    if (p.endsWith(".taglib.xml")) {
                        String urlString = context.getResource(p).toExternalForm();
                        urlString = urlString.replaceAll(" ", "%20");
                        try {
                            urlsList.add(new URI(urlString));
                        } catch (URISyntaxException ex) {
                            throw new FacesException(ex);
                        }
                    }
                }
            }
            return urlsList;
        } catch (IOException ioe) {
            throw new FacesException("Error searching classpath from facelet-taglib documents", ioe);
        }

    }
```

The above piece of code as a familiar line of code:
```

 URL[] externalTaglibUrls = Classpath.search(Util.getCurrentLoader(this),
                                          "META-INF/",
                                          SUFFIX);
```

This is expensive. This is another 2 seconds of my deployment time and on the initialization of mojarra.
This is a bit of an unsolvable problem, because unfortunately, unlikes for the faces-config.xml file, there is historically no core name that people just use 99% of the time.
So everybody is doing mylibrary.taglib.xml.
And this is quite.

I would suggest here, that it would be urgent to adding to the hunt for the taglib.xml the need to have a good name by convention.
Such as taglib.xml without any prefix, and that Mojarra should have a strategy for hunting for this file similar to the stragy to hunt for the faces-config.xml.
That is, to do a:

Enumeration<URL> e = Util.getCurrentLoader(this).getResources("META-INF/taglib.xml");
To get an almost immediate result, and allow for the more generic and expensive hunto for files called whatever.taglib.xml at any META-INFO to be skipper.

Right now, while deploying an application with several dependencies, this get resources step is costing about two seconds and it is systematically shown in samples doing stuff like:

```
-- ----------------------------------------------
-- SAMPLE 34
-- ----------------------------------------------
at java.io.WinNTFileSystem.getBooleanAttributes(Native Method) 
at java.io.File.isDirectory(Unknown Source) 
at org.jboss.vfs.spi.RootFileSystem.isDirectory(RootFileSystem.java:107) 
at org.jboss.vfs.VirtualFile.isDirectory(VirtualFile.java:291) 
at org.jboss.vfs.VirtualJarInputStream.getNextJarEntry(VirtualJarInputStream.java:101) 
at org.jboss.vfs.VirtualJarInputStream.getNextEntry(VirtualJarInputStream.java:80) 
at com.sun.faces.facelets.util.Classpath.searchFromURL(Classpath.java:222) 
at com.sun.faces.facelets.util.Classpath.searchFromURL(Classpath.java:242) 
at com.sun.faces.facelets.util.Classpath.search(Classpath.java:156) 
at com.sun.faces.facelets.util.Classpath.search(Classpath.java:116)

 
at com.sun.faces.config.configprovider.MetaInfFaceletTaglibraryConfigProvider.getResources(MetaInfFaceletTaglibraryConfigProvider.java:93) 
at com.sun.faces.config.ConfigManager$URITask.call(ConfigManager.java:1362) 
at com.sun.faces.config.ConfigManager$URITask.call(ConfigManager.java:1331) 
at java.util.concurrent.FutureTask.run(Unknown Source) 
at com.sun.faces.config.ConfigManager.getConfigDocuments(ConfigManager.java:742) 
at com.sun.faces.config.ConfigManager.initialize(ConfigManager.java:441) 
at com.sun.faces.config.ConfigureListener.contextInitialized(ConfigureListener.java:227) 

-- same as above
at io.undertow.servlet.core.ApplicationListeners.contextInitialized(ApplicationListeners.java:187) 
at io.undertow.servlet.core.DeploymentManagerImpl$1.call(DeploymentManagerImpl.java:200) 
at io.undertow.servlet.core.DeploymentManagerImpl$1.call(DeploymentManagerImpl.java:171) 
at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:42) 
at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44) 
at io.undertow.servlet.core.DeploymentManagerImpl.deploy(DeploymentManagerImpl.java:234) 
at org.wildfly.extension.undertow.deployment.UndertowDeploymentService.startContext(UndertowDeploymentService.java:100) 
at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:82) 
at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) 
at java.util.concurrent.FutureTask.run(Unknown Source) 
at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) 
at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) 
at java.lang.Thread.run(Unknown Source) 
at org.jboss.threads.JBossThread.run(JBossThread.java:320)


```

I believe that this bootstrapping time for mojarra is too high.
It should be very easy for a small tunning to mojarra to allow us to half this bootstraping time on the faces-config.xml front by skipping the classpath search.
And harder to do for the taglibs, but also possible.
The taglib is ultimately a very bad problem, because even if Mojarra comes up with a quick search approach for the taglib.xml files without doing the classpath search, we all get burned by all the library depencies we have that today name their taglibs whatever they want with the right suffix, and that should have otherwise been guided to use a proper name  like faces-config.xml

But ok, still workable.
Well developed libraries like delta spike would very readily adhere to these new conventions.
And slower moving bad maintained libraries ... well those would not.

Thanks for taking a look.

**Comment by arjantijms** [comment]
Hi,

Thanks a bunch for the report. I just started looking at this and found some other places were initialisation induces a high cost as well, such as a massive (1.2MB) file (`jsf-ri-runtime.xml`) being parsed at every deployment, as well as the `com.sun.faces.config.processor.FactoryConfigProcessor.verifyFactoriesExist()`, where specifically the `javax.faces.render.RenderKitFactory` takes relatively long to initialise.

At the moment I'm testing with the smallest app possible, basically only this class:

```java
@Named
@RequestScoped
@FacesConfig
public class IndexBean {

	public String getHello() {
		return "hello";
	}
	
}
```

And a single Facelet file, so that gives a different picture of where the slowness is. I'll first look at this small/trivial app case, and will then move to the larger apps one.

Thanks again!

**Comment by 99sono** [comment]
No Problem,
I would also recommend that at some point a deployment is tested against an apploication with many many useless libraries.
Just to pump dependencies into the web-inf/lib folder, that have nothing to do with JSF. Apache commons util or thinks like that.

Because, ultimately, it might be nice to have an approach that has higher cost of initialization if one absolutely needs mojarra to do the heavy lifting of navigating over folder structures.
But that a door exists as well for applications comprised of properly mantained libraries that have the metadata files at orderly locations. That for these applicatins the mojarra startup cost should be really cheap - by allowing to disable the more expensive scanning algorithms.

Many thanks for having a look.
Kindest regards.

**Comment by 99sono** [comment]
Hi,

Please be aware that a bug is open to oracle as well on this Topic.
Mojarra is being initialized once jax-ws (@WebService) deployed on a WAR application.

What this means is, if you have a WAr application, with a faces-config.xml and you put in there 20 web services and try to deploy the application Weblogic, the deployment time will automatically baloon to 20 seconds or more

**Comment by 99sono** [comment]
Hi, 

Do you have an idea of when you might have been able to find a good solution for this bootsrapping cost? And will be made available on JSF 2.2.x? 

Many thanks for all the help.

**Comment by 99sono** [comment]
Hi, 
I am still waiting for some feedback on this.
Many thanks for any feedback.

**Comment by arjantijms** [comment]
I'm currently still working on it. I did a bunch of optimisations, but subsequently a number of tests started to fail. The init and especially factory code is quite hairy and sensitive, so I'm detangling that first, including some refactoring so I can better understand what's going on (i.e. the terms FactoryFinder and FactoryManager are both used for the same thing, making it hard to keep all that in mind)

**Comment by edburns** [comment]
Please see this important message regarding community contributions to
Mojarra.

https://javaee.groups.io/g/jsf-spec/message/30

Also, please consider joining that group, as that group has taken the
place of the old dev@javaserverfaces.java.net mailing list.

Thanks,

Ed Burns

**Comment by arjantijms** [comment]
@99sono 
>Well developed libraries like delta spike would very readily adhere to these new conventions.
And slower moving bad maintained libraries ... well those would not.

Regarding the `*-taglib.xml` file name; it's not that simple to use a single name by convention. The issue is that one file declares a namespace, e.g. `http://omnifaces.org/functions` and that libraries not rarely need to declare multiple namespaces for multiple things.

You'd probably need a new file format for this to be really effective, one with sections for namespaces, or alternatively enumerating the names in a single well defined file.

**Comment by 99sono** [comment]
Hi arjantijms, many thanks for the feedback.

I understand the point. If you have more than one namespace, then you need more than one taglib, and this would be true for several important jsf based libraries.
Well then, what about having the possibility of giving the deployed application a single file:
taglib-index.xml

That would replace the convention of default-taglib.xml.
And then either via system property, or somet other configuration mechanism, I could tell JSF.
When scanning for taglibs, ignore any *-taglib.xml that is not mentioned via taglib-index.xml.

And in the taglib-index.xml, I would then add a list of entries such as:
META-INF/namespace1-taglib.xml
...

In short, instead of assuming that there is a single taglib file, I assume that an appliction has the possibility to force me to use a taglib-index as only search apporach.
In this file, I will find all the classpath search strings to be used.
And libraries that come without, miss the bus of  this optimization and force the deployment to go over the more expensive hunting classpath search.

Kindest regards,
And many thanks for all the help.

**Comment by 99sono** [comment]
Another thing that might also be a not a bad idea, is to run the hunt to faces-config.xml and tag libraries in parallel.

Something like:
List<Runnable> huntingActionsThatCanRunInParallel = ..;
add(huntForFacesConfig)
add(huntForTagLibgFiles)
huntingActionsThatCanRunInParallel.parallelStream().forEach(currentRunnable -> currentRunnable.run())

And those two actions would be done simulateneously.
This might bring a nice performance boost.
